### PR TITLE
[3326][IMP] base_custom_filter

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[3326](https://www.quartile.co/web#id=3326&action=771&model=project.task&view_type=form&menu_id=505)
The second commit(26c8e7c7339f8fbb22fd798a9d4151d091ae13eb) should be worked but unfortunatelyit is not working because adding groups using etree is not supported(may be bug). That's why I changed the approach(see in third commit a91d5cb4a9836b9f7c51c115d2b74b06d34c8a45)

### CODE REVIEW

File: base_custom_filter/models/ir_filters.py [hunk 1]
```diff
@@ -14,6 +14,7 @@ def _selection_type(self):
             ("favorite", "Favorite"),
             ("filter", "Standard Filter"),
             ("groupby", "Standard Group By"),
+            ("search", "Standard Search"),
```
Comment (yostashiro) [inline base_custom_filter/models/ir_filters.py]: Shouldn't this be inserted in between `favorite` and `filter`?

File: base_custom_filter/views/ir_filters_views.xml [hunk 2]
```diff
@@ -19,7 +33,7 @@
             <field name="domain" position="attributes">
                 <attribute
                     name="attrs"
-                >{'invisible': [('type', '=', 'groupby')]}</attribute>
+                >{'invisible': [('type', '=', 'groupby')], 'required': [('type','!=', 'search')]}</attribute>
```
Comment (yostashiro) [inline base_custom_filter/views/ir_filters_views.xml]: I don't understand why this change is needed.

File: base_custom_filter/views/ir_filters_views.xml [hunk 3]
```diff
@@ -9,6 +9,20 @@
...
+                    domain="[('model', '=', model_id)]"
+                    attrs="{'required': [('type', '=', 'search')], 'invisible': [('type', '!=', 'search')]}"
+                />
+                <field
+                    name="filter_domain"
+                    attrs="{'invisible': [('type', '!=', 'search')]}"
+                />
+                <field
+                    name="group_ids"
+                    attrs="{'invisible': [('type', '!=', 'search')]}"
```
Comment (yostashiro) [inline base_custom_filter/views/ir_filters_views.xml]: It makes the design incoherent if we limit this field only to `search`. Make it available to `filter` and `groupby`, and adjust `get_view()` accordingly.

File: base_custom_filter/views/ir_filters_views.xml [hunk 4]
```diff
@@ -9,6 +9,20 @@
...
+                    widget="many2many_tags"
+                />
                 <field
                     name="group_id"
```
Comment (yostashiro) [inline base_custom_filter/views/ir_filters_views.xml:27]: There is no point showing this field when type is `search`.

File: base_custom_filter/models/ir_filters.py [hunk 5]
```diff
@@ -22,9 +23,15 @@ def _selection_type(self):
...
+    search_field_id = fields.Many2one(
+        comodel_name="ir.model.fields",
+        ondelete="cascade",
+    )
     groupby_field = fields.Many2one(
         comodel_name="ir.model.fields",
         string="Group By Field",
         ondelete="cascade",
     )
+    filter_domain = fields.Text()
```
Comment (yostashiro) [inline base_custom_filter/models/ir_filters.py]: Would be nice to have a help text with a simple example.

File: base_custom_filter/models/base.py [hunk 6]
```diff
@@ -54,6 +54,26 @@ def _add_grouped_groupby(self, res, custom_groupbys):
         res["arch"] = etree.tostring(arch)
         return res
 
+    @api.model
+    def _add_search_field(self, res, search_fields):
+        xml_arch = etree.fromstring(res["arch"])
+        for search_field in search_fields:
+            group_ids = self.env.user.groups_id.ids
+            search_field_group_ids = search_field.group_ids.ids
+            if any(group in group_ids for group in search_field_group_ids) or not search_field_group_ids:
```
Comment (yostashiro) [inline base_custom_filter/models/base.py]: Create a method for this part.
```python
    @api.model
    def _user_has_access_to_item(self, item):
        if not item.group_ids:
            return True
        return bool(set(self.env.user.groups_id.ids) & set(item.group_ids.ids))
```

And then call it like this.

```suggestion
            if not self._user_has_access_to_item(search_field):
                continue
```

Apply this to relevant parts of other methods.

**Review by yostashiro — CHANGES_REQUESTED** [review]
Can you please add simple tests and update readme as well.

File: base_custom_filter/views/ir_filters_views.xml [hunk 7]
```diff
@@ -9,6 +9,19 @@
...
+                    attrs="{'required': [('type', '=', 'search')], 'invisible': [('type', '!=', 'search')]}"
+                />
+                <field
+                    name="filter_domain"
+                    attrs="{'invisible': [('type', '!=', 'search')]}"
+                />
+                <field
+                    name="group_ids"
+                    widget="many2many_tags"
+                />
```
Comment (yostashiro) [inline base_custom_filter/views/ir_filters_views.xml]: Should be hidden for type `favorite`.

File: base_custom_filter/models/ir_filters.py [hunk 8]
```diff
@@ -22,9 +23,21 @@ def _selection_type(self):
...
         string="Group By Field",
         ondelete="cascade",
     )
+    filter_domain = fields.Text(help='''
+                                    Enter a filter domain expression that will be used
+                                    to filter records when a search is conducted.
+                                    This should be in a domain-style format. The expression
+                                    should be a list of tuples, each specifying a condition.
+                                    For example, '[('field_name', 'operator', value)]'.'''
+                                )
```
Comment (yostashiro) [inline base_custom_filter/models/ir_filters.py]: Fix the styling and avoid a long explanation. The audience should already know the concept of the filter domain. IMO, something like the following would be sufficient.

```
help="Enter a filter domain expression if necessary. Example: [('default_code', 'ilike', self)]"
```

File: base_custom_filter/readme/CONFIGURE.rst [hunk 9]
```diff
@@ -4,9 +4,12 @@
    * The **Model** for which you are defining the filter.
    * The **Type** depending on whether you want to add a filter or a grouping by
      by a field.
+   * The **Search Field** that will be added to the search.
+   * The **Filter Domain** that will be applied with the search.
    * The **Domain** that will be applied with the filter.
    * The **Group By Field** used to perform the group by.
    * The **Group** to have filters under the same separator.
+   * The **User Groups** that will be applied to the search for the visibility of users.
```
Comment (yostashiro) [inline base_custom_filter/readme/CONFIGURE.rst]: Can you re-organize the content by type here? It's now confusing as a type only uses certain set of fields.

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Comment by rinaldifirdaus** [comment]
@yostashiro 
Functional test: 
add following custom filter, search result works fine!
image

Code review:
LGTM and the description is more clear with the latest update.

kindly your approval to apply this improvement to the staging database. Thanks!

File: base_custom_filter/models/base.py [hunk 10]
```diff
@@ -54,6 +64,25 @@ def _add_grouped_groupby(self, res, custom_groupbys):
...
+        for search_field in search_fields:
+            if not self._user_has_access_to_item(search_field):
+                continue
+            new_field = etree.Element(
+                "field",
+                name=search_field.search_field_id.sudo().name,
+            )
+            if search_field.filter_domain:
+                new_field.set("filter_domain", search_field.filter_domain)
+
```
Comment (yostashiro) [inline base_custom_filter/models/base.py]:
```suggestion
```

File: base_custom_filter/readme/DESCRIPTION.rst [hunk 11]
```diff
@@ -1,2 +1,2 @@
-This module allows to define custom filters to be shown under the standard
+This module allows to define custom search in model's search view, custom filters to be shown under the standard
 filters and group by menus of a model's search view.
```
Comment (yostashiro) [inline base_custom_filter/readme/DESCRIPTION.rst]: Refine the sentence. It currently looks a bit broken.

File: base_custom_filter/tests/test_filters.py [hunk 12]
```diff
@@ -6,6 +6,19 @@ class Test(TransactionCase):
     @classmethod
     def setUpClass(cls, chart_template_ref=None):
```
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py]:
```suggestion
    def setUpClass(cls):
```

File: base_custom_filter/tests/test_filters.py [hunk 13]
```diff
@@ -6,6 +6,19 @@ class Test(TransactionCase):
...
         super().setUpClass()
+
+        filters_obj = cls.env["ir.filters"]
+        filters_group = Form(filters_obj)
```
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py]: Why `filters_group`?  I think this variable name is misleading.  `filter_form` would be more appropriate.

File: base_custom_filter/tests/test_filters.py
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py:1]: I find these tests a bit random (e.g. the test name is test_sale_order_line() when it has nothing to do with sale order line) and of little value.

Can you add some patterns to check the content of the value returned from `get_view()`.

File: base_custom_filter/models/ir_filters.py [hunk 14]
```diff
@@ -12,6 +12,7 @@ class IrFilters(models.Model):
     def _selection_type(self):
         return [
             ("favorite", "Favorite"),
+            ("search", "Standard Search"),
```
Comment (yostashiro) [inline base_custom_filter/models/ir_filters.py]: What does this `Standard` mean, by the way?

File: base_custom_filter/readme/CONFIGURE.rst [hunk 15]
```diff
@@ -1,12 +1,30 @@
...
+      * The **Group** to have filters under the same separator.
+      * The **User Groups** that will be applied to the filter for the visibility of users.
+
+   For creating Group By:
+
+      * The **Model** for which you are defining the filter.
+      * The **Type** depending on whether you want to add a filter or a grouping by a field.
+      * The **Group By Field** used to perform the group by.
+      * The **Group** to have filters under the same separator.
+      * The **User Groups** that will be applied to the groupby for the visibility of users.
```
Comment (yostashiro) [inline base_custom_filter/readme/CONFIGURE.rst]:
```suggestion
#. Create a record assiging model, type (search/filter/groupby) and necessary attributes.
    Available fields and corresponding attributes (in brackets) for each type are as follows:

   Search:

      * Search Field (``name``)
      * Filter Domain (``filter_domain``)
      * Domain (``domain``)
      * User Groups (``groups``)

   Filter:

      * Domain (``domain``)
      * User Groups (``groups``)

   Group By:

      * Group By Field (field to be assigned to ``group_by`` context)
      * User Groups (``groups``)

   See `the official documentation <https://www.odoo.com/documentation/16.0/developer/reference/backend/views.html#search>`_ for the definition of each attribute.

   Additionally, filter and group-by records can be respectively grouped together with
   "Group" assignment (there will be a separator in between groups).
``

**Review by yostashiro — CHANGES_REQUESTED** [review]
@AungKoKoLin1997 Please go ahead and create a PR in the OCA repo so that the community knows what's happening.

File: base_custom_filter/views/ir_filters_views.xml (hunk 4 above)
Comment (yostashiro) [reply base_custom_filter/views/ir_filters_views.xml:27]: Please take care of this point.

File: base_custom_filter/tests/test_filters.py
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py:1]: @AungKoKoLin1997 Can you please follow up on the following:

- Normalize the setup and tests - prepare variables in the setup and user them in tests.
- Split test_base_custom_filter() into smaller pieces as it's currently a bit bloated. We may prepare one test per type, for example.

File: base_custom_filter/tests/test_filters.py [hunk 16]
```diff
@@ -4,8 +4,21 @@
...
+        filters_form = Form(filters_obj)
+        filters_form.name = "Test No Search group"
+        filters_form.type = "search"
+        filters_form.model_id = "ir.filters.group"
+        filters_form.search_field_id = cls.env.ref(
+            "base_custom_filter.field_ir_filters_group__display_name"
+        )
+        filters_form.filter_domain = "['display_name', 'ilike', self]"
+        filters_form.group_ids.add(cls.env.ref("base.group_system"))
+        filters_form = filters_form.save()
```
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py]: I'd prefer not to have these lines unless it's for the purpose of a "setup", which is currently not the case. IMO, there is no point adding another pattern to confirm that the form can be saved.

**Review by yostashiro — COMMENTED** [review]
Thanks for the update, Lin.

File: base_custom_filter/tests/test_filters.py [hunk 17]
```diff
@@ -16,30 +17,66 @@ def setUpClass(cls, chart_template_ref=None):
...
-    def test_sale_order_line(self):
-        filters_group_obj = self.env["ir.filters.group"]
-        filters_obj = self.env["ir.filters"]
-        filters_obj.unlink()
-        filters_group_obj.unlink()
-        with Form(filters_group_obj) as filters_group:
+    def test_get_view_content_search(self):
+        with Form(self.filters_group_obj) as filters_group:
+            filters_group.name = "Test Search"
+            filters_group.type = "search"
```
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py]: @AungKoKoLin1997 This pattern shouldn't exist according to your last update...

File: base_custom_filter/tests/test_filters.py [hunk 18]
```diff
@@ -16,30 +17,64 @@ def setUpClass(cls, chart_template_ref=None):
...
+        )
+        filter_string = '<filter name="{}" string="Test filter line" domain="[[&quot;id&quot;,&quot;=&quot;,1]]"/>'.format(  # noqa: B950
+            filter_name
+        )
+        self.assertIn(
+            filter_string,
+            view_content,
+            "The string is not in the returned view",
+        )
+        filter_group.unlink()
```
Comment (yostashiro) [inline base_custom_filter/tests/test_filters.py]: What is this `unlink()` for?

**Review by yostashiro — APPROVED** [review]

**Review by kanda999 — APPROVED** [review]

**Pull request merged**